ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,766 @@
+           subarray description
+        data : DL1CameraContainer
+            calibrated camera data
+        tel_id: int
+            which telescope is being cleaned
+
+        Returns
+        -------
+        np.ndarray, ImageParametersContainer:
+            cleaning mask, parameters
+        """
+
+        tel = subarray.tel[tel_id]
+        geometry = tel.camera.geometry
+
+        # apply cleaning
+        signal_pixels = self.clean(tel_id=tel_id, image=data.image)
+        image_selected = data.image[signal_pixels]
+
+        params = ExtendedImageParametersContainer()
+
+        # check if image can be parameterized:
+        image_criteria = self.check_image(image_selected)
+        self.log.debug(
+            "image_criteria: %s",
+            list(zip(self.check_image.criteria_names[1:], image_criteria)),
+        )
+
+        # parameterize the event if all criteria pass:
+        if all(image_criteria):
+            geom_selected = geometry[signal_pixels]
+
+            params.hillas = hillas_parameters(
+                geom=geom_selected, image=image_selected,
+            )
+            params.timing = timing_parameters(
+                geom=geom_selected,
+                image=image_selected,
+                pulse_time=data.pulse_time[signal_pixels],
+                hillas_parameters=params.hillas,
+            )
+            params.leakage = leakage(
+                geom=geometry, image=data.image, cleaning_mask=signal_pixels
+            )
+            params.concentration = concentration(
+                geom=geom_selected,
+                image=image_selected,
+                hillas_parameters=params.hillas,
+            )
+            params.morphology = morphology(
+                geom=geometry, image_mask=signal_pixels
+            )
+            params.intensity = intensity_statistics(image=image_selected)
+
+        return signal_pixels, params
+
+    def _process_events(self, writer):
+        self.log.debug("Writing DL1/Event data")
+        tel_index = TelEventIndexContainer()
+        event_index = EventIndexContainer()
+        is_initialized = False
+        self.event_source.subarray.info(printer=self.log.debug)
+
+        for event in tqdm(
+            self.event_source,
+            desc=self.event_source.__class__.__name__,
+            total=self.event_source.max_events,
+            unit="ev",
+            disable=not self.progress_bar,
+        ):
+
+            self.log.log(9, "Writing event_id=%s", event.dl0.event_id)
+
+            self.calibrate(event)
+
+            event.mc.prefix = "mc"
+            event.trig.prefix = ""
+            event_index.event_id = event.index.event_id
+            event_index.obs_id = event.index.obs_id
+            tel_index.event_id = event.index.event_id
+            tel_index.obs_id = event.index.obs_id
+            self._cur_obs_id = event.index.obs_id
+
+            # On the first event, we now have a subarray loaded, and other info, so
```

Comment:
subarray is loaded before the first event!

Reply:
remnant from before that was implemented. Will remove

Reply:
ALready fixed it in my PR

Reply:
https://github.com/kosack/ctapipe/pull/8